pkg/controller/quota/quota.go

```diff
@@ -0,0 +1,128 @@
+	// Grab the config and check that the quota system is enabled
+	cfg, err := config.Get(req.Ctx, req.Client)
+	if err != nil {
+		return err
+	}
+	if !*cfg.QuotaEnabled {
+		return nil
+	}
+
+	appInstance := req.Object.(*apiv1.AppInstance)
+	status := condition.Setter(appInstance, resp, apiv1.AppInstanceConditionQuotaAllocated)
+
+	// Create the quota request object and give calculate the standard numeric values
+	name, namespace, app := appInstance.Name, appInstance.Status.Namespace, appInstance.Status.AppSpec
+	quotaRequest := &adminv1.QuotaRequestInstance{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace},
+		Spec: adminv1.QuotaRequestInstanceSpec{
+			Containers: len(app.Containers),
+			Jobs:       len(app.Jobs),
+			Volumes:    len(app.Volumes),
+			Secrets:    len(app.Secrets),
+			Images:     len(app.Images),
+
+			// Create default quantity values to be added to
+			VolumeStorage: *resource.NewQuantity(0, resource.BinarySI),
+			CPU:           *resource.NewMilliQuantity(0, resource.BinarySI),
+			Memory:        *resource.NewQuantity(0, resource.BinarySI),
+			RootStorage:   *resource.NewQuantity(0, resource.BinarySI),
+		},
+	}
+
+	// Add the more complex values to the quota request
+	addCompute(app.Containers, appInstance, quotaRequest)
+	addCompute(app.Jobs, appInstance, quotaRequest)
+	if err := addStorage(appInstance, quotaRequest); err != nil {
+		status.Error(err)
+		return err
+	}
+
+	resp.Objects(quotaRequest)
+	return nil
+}
+
+// addCompute adds the compute resources of the containers passed to the quota request.
+func addCompute(containers map[string]apiv1.Container, appInstance *apiv1.AppInstance, quotaRequest *adminv1.QuotaRequestInstance) {
+	// For each workload, add their memory/cpu requests to the quota request
+	for name, container := range containers {
+		if _, ok := appInstance.Status.Scheduling[name]; ok {
+			quotaRequest.Spec.CPU.Add(appInstance.Status.Scheduling[name].Requirements.Requests["cpu"])
+			quotaRequest.Spec.Memory.Add(appInstance.Status.Scheduling[name].Requirements.Requests["memory"])
+		} else if _, ok := appInstance.Status.Scheduling[""]; ok {
+			quotaRequest.Spec.CPU.Add(appInstance.Status.Scheduling[name].Requirements.Requests["cpu"])
+			quotaRequest.Spec.Memory.Add(appInstance.Status.Scheduling[name].Requirements.Requests["memory"])
+		}
+		// Recurse over any sidecars. Since sidecars can't have sidecars, this is safe.
+		addCompute(container.Sidecars, appInstance, quotaRequest)
+	}
+}
+
+// addStorage adds the storage resources of the volumes passed to the quota request.
+func addStorage(appInstance *apiv1.AppInstance, quotaRequest *adminv1.QuotaRequestInstance) error {
+	app := appInstance.Status.AppSpec
+
+	// Add the volume storage needed to the quota request.
+	for _, volume := range app.Volumes {
+		size, err := resource.ParseQuantity(string(volume.Size))
+		if err != nil {
+			return err
+		}
+		quotaRequest.Spec.VolumeStorage.Add(size)
+	}
```

Comment:
Since this is grabbing `Volumes` from the AppSpec, this will only calculate storage for ones that are specified in the Acornfile. This is an issue as it means that `Volumes` specified client side in the `AppInstance` can effectively get around being tracked in the quota. I would just process the `AppInstance` `Volumes` directly but that implements a second source of truth for the state of `Volumes`.

Reply:
Actually this isn't a problem as the `Volumes` defined in the `AppInstance` only overwrites existing `Volumes` in the `Acornfile`.